### PR TITLE
spar mt prompt invar

### LINKED ISSUE #172: Combine prompts to evaluate multilingual prompt invariance
- add cmdline option --combined_prompts_path=<path_to_save_combined_template>
  - check that dataset col number and names match
  - if so, combine into single DatasetsTemplate and use in load_prompts loop

### PR DESCRIPTION
**Pull request by ChristyKoh**
Fixes #172 
- add merge_templates_from method to DatasetTemplates
- generate combined templates.yaml when PromptConfig is initialized

currently, checks for alignment of col number and ClassLabel classes

### CODE REVIEW

File: elk/extraction/prompt_loading.py [hunk 1]
```diff
@@ -69,6 +74,25 @@ def __post_init__(self):
...
+                + f"{self.combined_template_path}/templates.yaml"
+            )
+            combined_prompter = DatasetTemplates(
+                "combined_templates", self.combined_template_path
+            )
+            combined_prompter.templates = {}
+            for ds_string in self.datasets:
+                ds_name, _, config_name = ds_string.partition(" ")
+                prompter = DatasetTemplates(ds_name, config_name)
+                # TODO: Verify that cols are same; if not, warn that templates
```
Comment (lauritowal) [inline elk/extraction/prompt_loading.py:90]: we should add that

**Review by lauritowal — CHANGES_REQUESTED** [review]

File: elk/extraction/prompt_loading.py [hunk 2]
```diff
@@ -136,12 +164,20 @@ def load_prompts(
...
+        prompters = [combined_prompter] * len(dataset_strings)
+
     min_num_templates = min(len(prompter.templates) for prompter in prompters)
+
     num_variants = (
         min_num_templates
         if num_variants == -1
         else min(num_variants, min_num_templates)
     )
+    print()
```
[the remark this answers is not in the document]
Comment (ChristyKoh) [reply elk/extraction/prompt_loading.py]: Done.

File: elk/extraction/prompt_loading.py (hunk 1 above)
Comment (ChristyKoh) [reply elk/extraction/prompt_loading.py]: Done.

File: elk/extraction/extraction.py [hunk 3]
```diff
@@ -268,8 +269,10 @@ def get_splits() -> SplitDict:
     model_cfg = AutoConfig.from_pretrained(cfg.model)
     num_variants = cfg.prompts.num_variants
 
+    # extraneous, remove ?
```
[the remark this answers is not in the document]
Comment (ChristyKoh) [reply elk/extraction/extraction.py]: My bad, I realized they're used in the get_splits() function above. Added a comment.

File: elk/extraction/prompt_loading.py [hunk 4]
```diff
@@ -69,6 +75,60 @@ def __post_init__(self):
         if len(self.max_examples) == 1:
             self.max_examples *= 2
 
+        # Combining prompts
```
Comment (lauritowal) [inline elk/extraction/prompt_loading.py:78]: Let's maybe split this into multiple functions

File: elk/extraction/prompt_loading.py [hunk 5]
```diff
@@ -69,6 +75,60 @@ def __post_init__(self):
...
+                    )
+                    self.combined_template_path = ""
+                    break
+                prev_num_label_classes = num_classes
+
+                # Once verified, merge templates.
+                combined_prompter.merge_templates_from(prompter)
+
+        # Write to file if successfully merged all prompts.
+        if self.combined_template_path:
```
Comment (lauritowal) [inline elk/extraction/prompt_loading.py:123]: You can remove this, since you are checking above already

Comment (lauritowal) [reply elk/extraction/prompt_loading.py:123]: but yeah, this can also be an own function

File: elk/extraction/prompt_loading.py [hunk 6]
```diff
@@ -41,9 +42,13 @@ class PromptConfig(Serializable):
         num_shots: The number of examples to use in few-shot prompts. If zero, prompts
             are zero-shot. Defaults to 0.
         num_variants: The number of prompt templates to apply to each predicate upon
-            call to __getitem__. Use -1 to apply all available templates. Defaults to 1.
+            call to __getitem__. Use -1 to apply all available templates. Defaults to
+            -1.
         seed: The seed to use for prompt randomization. Defaults to 42.
         stream: Whether to stream the dataset from the Internet. Defaults to False.
+        combined_template_path: Path to save a combined template file to, when testing
```
Comment (lauritowal) [inline elk/extraction/prompt_loading.py:49]: maybe combined_template_output_path describes what it is better.

Comment (lauritowal) [inline elk/extraction/prompt_loading.py:49]: when testing --> when applying or when using ?

File: elk/extraction/prompt_loading.py [hunk 7]
```diff
@@ -105,7 +166,10 @@ def load_prompts(
     # templates for any dataset in order to make sure we don't run out of prompts.
     for ds_string in dataset_strings:
         ds_name, _, config_name = ds_string.partition(" ")
-        prompters.append(DatasetTemplates(ds_name, config_name))
+
+        if combined_template_path == "":
```
Comment (lauritowal) [inline elk/extraction/prompt_loading.py:170]: nitpick, but I guess 

`if not combined_template_path: `

is a little bit cleaner

**Review by lauritowal — CHANGES_REQUESTED** [review]
Thanks for the pull-request! Just tried it out and it seems to work. You might want to split up your code into multiple functions to make it a bit more readable. I've added some comments.

**Review by lauritowal — APPROVED** [review]

**Pull request merged**